### PR TITLE
mp4 (bit dirty, incomplete)

### PR DESCRIPTION
**Pull request by petr-briza**
Parser for MP4 container based on ISO/IEC 14496-14:2003 specs. There are some comments at the start of .ksy. Not all of "box" (format atomic blocks) types are implemented but enough to get some media metadata (like lenght, track types)

### CODE REVIEW

**Comment by GreyCat** [comment]
Thanks for this submission! Can I ask you to review the [style guide](http://doc.kaitai.io/ksy_style_guide.html) and apply some changes to this to match? I'll add more specific comments to the file itself.

File: media/mp4.ksy [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+#
+# -> moov -> mvhd -> duration / timescale    :: this is how you calculate media length in seconds
+# -> moov -> trak                            :: those are tracks in the container (video/audio/etc.)
+# trak -> tkhd -> @width x @height           :: resolution of the video track
+# trak -> mdia -> minf -> stbl -> stsd.entries[0].type :: fourcc of codec used for the track
+# trak -> mdia -> mdhd -> langCode           :: language code of the track
+#
+# trak -> mdia -> stb -> stts                :: table to get sample index for given time
+# trak -> mdia -> stb -> stsc                :: table to get chunk index for given time
+# trak -> mdia -> stb -> stco                :: table to get offset in file for given chunk index
```
Comment (GreyCat) [inline media/mp4.ksy]: This block of text looks like a good canididate for top-level `doc` element?

File: media/mp4.ksy [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+#
+# See also:
+# * http://mp4parser.com/
+# * https://archive.codeplex.com/?p=mp4explorer
+# * https://github.com/axiomatic-systems/Bento4
```
Comment (GreyCat) [inline media/mp4.ksy]: And these look like candidates for `doc-ref`?

File: media/mp4.ksy [hunk 3]
```diff
@@ -0,0 +1,408 @@
+meta:
+  id: mp4
+  file-extension: mp4
+  endian: be
+  license: CC0-1.0
```
Comment (GreyCat) [inline media/mp4.ksy]: At the very least, we'll need `title` here so it would be possible to find this file on the website.

File: media/mp4.ksy [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    type: box
+    repeat: eos
+
+types:
+
+  boxes:
+    seq:
+      - id: boxes
+        type: box
+        repeat: eos
```
Comment (GreyCat) [inline media/mp4.ksy]: This looks like the very same as root type, i.e. `mp4`?

File: media/mp4.ksy [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        type: str
+        size: 4
+      - id: largesize
+        type: u8
+        if: size == 1
+      - id: body
+        type:
+          switch-on: type
+          cases:
+            '"moov"': boxes
```
Comment (GreyCat) [inline media/mp4.ksy]: You can use root type, i.e. `mp4` instead of `boxes` everywhere.

File: media/mp4.ksy [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+        type: u4
+      - id: entry_count
+        type: u4
+      - id: entries
+        type: box_body_stsd_entry
+        repeat: expr
+        repeat-expr: entry_count
+        
+  box_body_stsd_entry:
+    seq:
```
Comment (GreyCat) [inline media/mp4.ksy]: For subtypes like these (i.e. which are clearly contained to one "box"), please consider using nested types, i.e.

```yaml
seq:
  # ...
  - id: entries
    type: entry
    repeat: expr
    repeat-expr: entry_count
types:
  entry:
# ...
```

File: media/mp4.ksy [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+            '"vmhd"': box_body_vmhd
+            '"mdhd"': box_body_mdhd
+            '"tkhd"': box_body_tkhd
+            '"elst"': box_body_elst
+            '"stsc"': box_body_stsc
+            '"stsd"': box_body_stsd
+            '"stco"': box_body_stco
+            '"stts"': box_body_stts
+            '"co64"': box_body_co64
+            _: box_body_other
```
Comment (GreyCat) [inline media/mp4.ksy]: You can just skip this as get it automatically captured as byte array if a match fails.

File: media/mp4.ksy [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+  boxes:
+    seq:
+      - id: boxes
+        type: box
+        repeat: eos
+
+  box:
+    seq:
+      - id: size
+        type: u4
```
Comment (GreyCat) [inline media/mp4.ksy]: `len_body` to match style guide?

**Comment by petr-briza** [comment]
Hi @GreyCat actually I noticed now that QuickTime .mov is almost the same format. Is there any possibility of "including" between several .ksy files? Cannot find it in doc. Maybe good idea would be to merge the same structures between those formats into one "included" file.

File: media/mp4.ksy [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+  # -----------
+  box_body_mvhd:
+    seq:
+      - id: version
+        type: u4
+
+      - id: creation_time
+        type:
+          switch-on: version
+          cases:
```
Comment (KOLANICH) [inline media/mp4.ksy]: I see pretty many records like tyis one, shoult tyis selector be encapsulated into a type?

**Comment by GreyCat** [comment]
@petr-briza 
> actually I noticed now that QuickTime .mov is almost the same format.

Yeah, I've just realized that too. Actually, I have a very stupid question: are they actually different file formats, and is there any value in keeping them separate?

> Is there any possibility of "including" between several .ksy files? Cannot find it in doc. Maybe good idea would be to merge the same structures between those formats into one "included" file.

Sure, we have [imports](http://doc.kaitai.io/user_guide.html#_importing_types_from_other_files), and, in 0.9 you can even reuse internal structures of the other file. But before jumping to that, let's think if we *really* want to have them separate?

**Comment by petr-briza** [comment]
@GreyCat

> But before jumping to that, let's think if we _really_ want to have them separate?

Actuallly I think: yes. For the very same reason I didn't noticed it's the same format as QuickTime .mov – people might now know it's the same format. If you want to keep only single .ksy file, what options do you have for the file name?

- Something like _ISO-14496-12-2015.ksy_ – most precise but most confusing. Who is no expert on the format (yet) might just not know that this ISO means .mov or .mp4. Such developer will just look to the list of formats, he won's see the desired format and move on (maybe will start writing his own parser, just like me).
- You can have both file formats in the name, like _mp4_mov_iso.ksy_ but this is messy, cannot be sorted in alphabetical order
- My suggestion is just to have two separate .ksy files, one for .mov and one for .mp4, and inside you would have "imports" for the shared part – which could be entire definition, int his case. MP4 is extension of the ISO, I just didn't reach level of implementation yet where I actually used any of the extended boxes.

You might also noticed there is the same problem for example with Microsoft PE format – which is used for .exe, .dll, .sys but also .ax, .drv, .src etc. If somebody wants to parse .ax to find a company name in the file etc, he might not realize there is already a parser called microsoft_pe.ksy.

Another approach would be to create "index" of parsers where you would have a table of "file extensions" → parser, another could be "mime type" → parser. This would help to find a parser easily and in such cases I would recommend to go for the "precise" file names like "ISO-14496-12-2015.ksy"

In all cases I would also recommend separating "file parsers" from "network protocol parsers", "memory block" whatever, etc. For other categories you could have different "index tables" based on different keys.

What do you think?

**Comment by GreyCat** [comment]
Ok, so there are two separate topics:

* Is there any real difference between these two formats that warrants separation? For example, if one clearly includes things which should be treated differently in the other?
* How to make these formats easier to find if people are searching for "mov", "mp4" and other popular terms?

I believe the 2nd question is more universal and is not specific to this particular case (as you've already demonstrated with .exe, .dll, .ax etc examples). Let's focus on the first issue, get our fundamentals perfect and then we'll improve navigation/search experience separately.

**Comment by petr-briza** [comment]
Answer to the second question will have influence on decisions for the first question, because:

- If you want to keep the names according to firmat standard itself, you should rename the "QuickTime MOV" to more generic "ISO" name. And the 1st answer would be "no, this should be just one format". At least at the current implementation stage when we don't have any .MOV-specific or any .MP4-specific boxes implemented yet.
- If you want naming system rarther according to "file extensions" or "mime types", in such case I would answer to the 1st question "yes, let's separate the file formats, call them MP4 and MOV but le's put the shared structures to the shared place"

In both cases you would end up with zero(or almost zero) code duplicity. But you should answer second question first.

**Comment by petr-briza** [comment]
Btw this is what we are talking about: https://en.wikipedia.org/wiki/ISO/IEC_base_media_file_format
I based my parser on the last edition of the spec and as you can see, it's actually based on original .mov and a bunch of format are extensions of this format: MP4, 3GP, 3G2, .mj2, .dvb, .dcf, .m21, .f4v. Of course  mp4 is most common.

**Comment by GreyCat** [comment]
Considering everything above, my proposal is then to:

- [ ] Rename current `quicktime_mov` to `iso_iec_base_media`
- [ ] Add documentation wording that will make it super clear that "mp4", "quicktime", "3gp", etc, are all the same format.
- [ ] Add relevant file extensions to `meta/file-extensions`
- [ ] Merge whatever is implemented in this PR and is missing in `iso_iec_base_media` there.
- [ ] Start a [big format-style](https://github.com/kaitai-io/kaitai_struct/issues/165) repo for this project, move the development there
- [ ] Enhance formats.kaitai.io to make sure it actually has way to find a file format by extension and catches all the popular asks like "mp4 format", "3gp format", etc.

**Comment by petr-briza** [comment]
OK I just found a time to move this forward. I don't think there is a reason to create separate repositories for "big formats" and also I don't think this is such a big format. So I just merged my improvements to the original "mov" format, renamed it as you recommended and created a new pull request: https://github.com/kaitai-io/kaitai_struct_formats/pull/272

**Pull request closed without merging**